File: leetcode/93.restore-ip-addresses/go/restoreIpaddresses.go
package restoreip

import (
	"strconv"
	"strings"
)

func restoreIpAddresses(s string) []string {
	res := make([]string, 0)
	path := make([]string, 0)

	var backtrack func(start int)

	backtrack = func(start int) {
		if start >= len(s) && len(path) == 4 {
			ip := strings.Join(path, ".")
			res = append(res, ip)
			return
		}

		for i := start; i < len(s); i++ {
			if len(path) >= 4 {
				break
			}

			part := s[start : i+1]
			if !isValidIpPart(part) {
				continue
			}

			path = append(path, part)
			backtrack(i + 1)
			path = path[:len(path)-1]
		}
	}

	backtrack(0)
	return res
}

func isValidIpPart(s string) bool {
	if len(s) > 1 && s[0] == '0' {
		return false
	}

	if len(s) > 3 {
		return false
	}

	i, err := strconv.Atoi(s)
	if err != nil {
		return false
	}

	return i >= 0 && i <= 255
}

File: leetcode/93.restore-ip-addresses/ts/restoreipaddresses.ts
const restoreIpAddresses = (s: string): string[] => {
    const res: string[] = []
    const path: string[] = []

    const backtrack = (start: number) => {
        if (path.length === 4 && start === s.length) {
            res.push(path.join("."))
            return
        }

        for (let i = start; i < s.length; i++) {
            const part = s.slice(start, i + 1)
            if (!isValidIpPart(part)) break

            path.push(part)
            backtrack(i + 1)
            path.pop()
        }
    }

    backtrack(0)
    return res
}

const isValidIpPart = (s: string): boolean => {
    if (s.length > 1 && s[0] === "0") return false

    if (s.length > 3) return false

    const part = parseInt(s)
    if (Number.isNaN(part)) return false

    return part >= 0 && part <= 255
}

export default restoreIpAddresses
